File: app/src/main/java/com/example/solar_alarm/Data/Tables/SolarTime.java
package com.example.solar_alarm.Data.Tables;

import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Index;

import com.example.solar_alarm.Data.Enums.SolarTimeTypeEnum;
import com.example.solar_alarm.sunrise_sunset_http.SunriseSunsetResponse;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

@RequiresApi(api = Build.VERSION_CODES.O)
@Entity
(
    tableName = "SolarTime",
    indices = {@Index(value = {"Date", "LocationId"}, unique = true)}
)

public class SolarTime extends TableBase
{
    @NonNull
    public LocalDate Date;

    @ForeignKey(entity = Location.class, parentColumns = "Id", childColumns = "LocationId")
    public int LocationId;
    public int DayLength;

    // UTC Time Stings in ISO_OFFSET_DATE_TIME format
    public String SunriseUtc;
    public String SunsetUtc;
    public String SolarNoonUtc;
    public String CivilTwilightBeginUtc;
    public String CivilTwilightEndUtc;
    public String NauticalTwilightBeginUtc;
    public String NauticalTwilightEndUtc;
    public String AstronomicalTwilightBeginUtc;
    public String AstronomicalTwilightEndUtc;

    public SolarTime (){}   // explicitly create default constructor for Android Room

    @RequiresApi(api = Build.VERSION_CODES.O)
    public SolarTime (Location location, SunriseSunsetResponse sunriseSunsetResponse)
    {
        Date                         = LocalDateTime.ofInstant(sunriseSunsetResponse.request.Date.toInstant(), ZoneId.systemDefault()).toLocalDate();
        LocationId                   = location.Id;
        DayLength                    = sunriseSunsetResponse.getDayLength();
        SunriseUtc                   = sunriseSunsetResponse.getSunrise();
        SunsetUtc                    = sunriseSunsetResponse.getSunset();
        SolarNoonUtc                 = sunriseSunsetResponse.getSolarNoon();
        CivilTwilightBeginUtc        = sunriseSunsetResponse.getCivilTwilightBegin();
        CivilTwilightEndUtc          = sunriseSunsetResponse.getCivilTwilightEnd();
        NauticalTwilightBeginUtc     = sunriseSunsetResponse.getNauticalTwilightBegin();
        NauticalTwilightEndUtc       = sunriseSunsetResponse.getNauticalTwilightEnd();
        AstronomicalTwilightBeginUtc = sunriseSunsetResponse.getAstronomicalTwilightBegin();
        AstronomicalTwilightEndUtc   = sunriseSunsetResponse.getAstronomicalTwilightEnd();
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    public ZonedDateTime GetLocalZonedDateTime(SolarTimeTypeEnum solarTimeTypeEnum) throws Exception
    {
        ZonedDateTime utcDateTime   = GetUtcZonedDateTime(solarTimeTypeEnum);
        ZoneId        zoneId        = ZoneId.systemDefault();
        ZonedDateTime LocalDateTime = utcDateTime.withZoneSameInstant(zoneId);

        return LocalDateTime;
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private ZonedDateTime GetUtcZonedDateTime(SolarTimeTypeEnum solarTimeTypeEnum) throws Exception
    {
        LocalDateTime localDateTime = getLocalDateTime(solarTimeTypeEnum);
        ZoneId        zoneId        = ZoneId.ofOffset("UTC", ZoneOffset.UTC);
        ZonedDateTime zonedDateTime = localDateTime.atZone(zoneId);

        return zonedDateTime;
    }

    private LocalDateTime getLocalDateTime(SolarTimeTypeEnum solarTimeTypeEnum) throws Exception
    {
        String utcString;

        switch (solarTimeTypeEnum)
        {
            case Sunrise:
                utcString = SunriseUtc;
                break;
            case Sunset:
                utcString = SunsetUtc;
                break;
            case SolarNoon:
                utcString = SolarNoonUtc;
                break;
            case CivilTwilightBegin:
                utcString = CivilTwilightBeginUtc;
                break;
            case CivilTwilightEnd:
                utcString = CivilTwilightEndUtc;
                break;
            case NauticalTwilightBegin:
                utcString = NauticalTwilightBeginUtc;
                break;
            case NauticalTwilightEnd:
                utcString = NauticalTwilightEndUtc;
                break;
            case AstronomicalTwilightBegin:
                utcString = AstronomicalTwilightBeginUtc;
                break;
            case AstronomicalTwilightEnd:
                utcString = AstronomicalTwilightEndUtc;
                break;
            default:
                throw new Exception("Enum not implemented");
        }

        LocalDateTime localDateTime = LocalDateTime.parse(utcString, DateTimeFormatter.ISO_OFFSET_DATE_TIME);

        return localDateTime;
    }
}

File: app/src/main/java/com/example/solar_alarm/Data/Tables/SolarTime.kt
package com.example.solar_alarm.Data.Tables

import android.os.Build
import android.os.Parcelable
import androidx.annotation.RequiresApi
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import com.example.solar_alarm.Data.Enums.SolarTimeTypeEnum
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter

@RequiresApi(api = Build.VERSION_CODES.O)
@Entity
(
        tableName   = "SolarTime",
        indices     =
        [
            Index(value = ["SolarDate", "LocationId"], unique = true,  name = "UniqueDateAndLocation"),
            Index(value = ["LocationId"],              unique = false, name = "SolarTimeLocationIndex")
        ],
        foreignKeys =
        [
            ForeignKey
            (
                entity        = Location::class,
                parentColumns = ["Id"],
                childColumns  = ["LocationId"],
                onDelete      = ForeignKey.CASCADE
            )
        ]
)

@Parcelize
data class SolarTime
(
    @ColumnInfo(name = "SolarDate")                    val SolarDate: LocalDate,
    @ColumnInfo(name = "LocationId")                   val LocationId: Int,
    @ColumnInfo(name = "DayLength")                    val DayLength: Int,
    @ColumnInfo(name = "SunriseUtc")                   val SunriseUtc: String?,
    @ColumnInfo(name = "SunsetUtc")                    val SunsetUtc: String?,
    @ColumnInfo(name = "SolarNoonUtc")                 val SolarNoonUtc: String?,
    @ColumnInfo(name = "CivilTwilightBeginUtc")        val CivilTwilightBeginUtc: String?,
    @ColumnInfo(name = "CivilTwilightEndUtc")          val CivilTwilightEndUtc: String?,
    @ColumnInfo(name = "NauticalTwilightBeginUtc")     val NauticalTwilightBeginUtc: String?,
    @ColumnInfo(name = "NauticalTwilightEndUtc")       val NauticalTwilightEndUtc: String?,
    @ColumnInfo(name = "AstronomicalTwilightBeginUtc") val AstronomicalTwilightBeginUtc: String?,
    @ColumnInfo(name = "AstronomicalTwilightEndUtc")   val AstronomicalTwilightEndUtc: String?
) : Parcelable
{
    @IgnoredOnParcel
    @PrimaryKey(autoGenerate = true) @ColumnInfo(name = "Id") var Id : Int = 0
    @IgnoredOnParcel
    @ColumnInfo(name = "CreateDateTimeUtc") var CreateDateTimeUtc : OffsetDateTime = OffsetDateTime.of(OffsetDateTime.now().toLocalDateTime(), ZoneOffset.UTC)

    /**
     * @param solarTimeTypeEnum the type of time.
     * @return UTC date time string in UTC time zone as a ZonedDateTime
     * @see ZonedDateTime
     */
    fun GetLocalZonedDateTime(solarTimeTypeEnum: SolarTimeTypeEnum): ZonedDateTime
    {
        val utcDateTime = GetUtcZonedDateTime(solarTimeTypeEnum)
        val zoneId      = ZoneId.systemDefault()

        return utcDateTime.withZoneSameInstant(zoneId)
    }

    /**
     * @param solarTimeTypeEnum the type of time.
     * @return UTC date time string in UTC time zone as a ZonedDateTime
     * @see ZonedDateTime
     */
    private fun GetUtcZonedDateTime(solarTimeTypeEnum: SolarTimeTypeEnum): ZonedDateTime
    {
        val localDateTime = getLocalDateTime(solarTimeTypeEnum)
        val zoneId        = ZoneId.ofOffset("UTC", ZoneOffset.UTC)

        return localDateTime.atZone(zoneId)
    }

    /**
     * @param solarTimeTypeEnum the type of time.
     * @return UTC date time string converted to the device's time zone as a LocalDateTime
     * @see LocalDateTime
     */
    private fun getLocalDateTime(solarTimeTypeEnum: SolarTimeTypeEnum): LocalDateTime
    {
        val utcString: String? = when (solarTimeTypeEnum)
        {
            SolarTimeTypeEnum.Sunrise                   -> SunriseUtc
            SolarTimeTypeEnum.Sunset                    -> SunsetUtc
            SolarTimeTypeEnum.SolarNoon                 -> SolarNoonUtc
            SolarTimeTypeEnum.CivilTwilightBegin        -> CivilTwilightBeginUtc
            SolarTimeTypeEnum.CivilTwilightEnd          -> CivilTwilightEndUtc
            SolarTimeTypeEnum.NauticalTwilightBegin     -> NauticalTwilightBeginUtc
            SolarTimeTypeEnum.NauticalTwilightEnd       -> NauticalTwilightEndUtc
            SolarTimeTypeEnum.AstronomicalTwilightBegin -> AstronomicalTwilightBeginUtc
            SolarTimeTypeEnum.AstronomicalTwilightEnd   -> AstronomicalTwilightEndUtc
        }

        return LocalDateTime.parse(utcString, DateTimeFormatter.ISO_OFFSET_DATE_TIME)
    }
}
